elastic/kibana
# Hides unknown uiSettings from the advanced settings page

Resolves https://github.com/elastic/kibana/issues/94876

Settings that aren't registered by plugins are visible and editable in Stack Management > Kibana > Advanced Settings > General. We need to hide these from the UI without removing them from the underlying saved object. Hiding `custom` settings will also hide any settings registered by a plugin that's been disabled (e.g Reporting).

Rather than introduce another flag, we reuse the [isCustom](https://github.com/elastic/kibana/blob/main/src/core/public/ui_settings/types.ts#L74-L79) property.

In this PR, 
- this PR filters out settings for disabled plugins by using the ui_settings client `isCustom` flag to decide on when to show a setting or not.
- we’re effectively ignoring previously set items that were for a plugin that is now disabled by filtering these out from the `AdvancedSettings` page.

This PR doesn't modify the underlying `uiSettings` saved object directly, because that should be done using an explicit migration.

## How to test this:
- Run Kibana with a trial license with all plugins enabled
- Navigate to Advanced Settings and adjust the settings for any of the plugins that may still be disabled. For example, add a custom pdf logo for reporting.
- Stop Kibana but **keep elasticsearch running** (retains previous settings in the `config` saved object)
- Disable any of the plugins that can be disabled, for example, to disable reporting:
```Kibana.yml
xpack.reporting.enabled: false
xpack.reporting.queue.pollEnabled: false
```
- Start Kibana again
- Navigate to Advanced Settings and search for any setting applicable to the plugin(s) you’ve disabled.
- Observe that the UI doesn’t render unregistered settings (even though they are in the config).
- You can confirm that the saved object still exists and is persisted by checking that settings for the disabled plugin are still in the config document. For example, with the reporting test setup:
```
GET .kibana/_search
{
  "query": {
    "bool": {
      "must": [
        {
          "exists": {
            "field": "config"
          }
        }
      ]
    }
  }
}
```
Should return a document similar to:
```
"config" : {
   "buildNum" : 9007199254740991,
   "xpackReporting:customPdfLogo" : data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAUAAAAFACA……
}
```
The [following plugins](https://github.com/elastic/kibana/issues/89584#issuecomment-943750618) can still be disabled:
- interactive_setup
- newsfeed
- reporting
- telemetry
- vis_type_* plugins

### Checklist
### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process) (no changes)
